cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,280 @@
+import co.cask.cdap.data2.queue.QueueClientFactory;
+import co.cask.cdap.data2.transaction.Transactions;
+import co.cask.cdap.data2.transaction.queue.ConsumerEntryState;
+import co.cask.cdap.data2.transaction.queue.QueueAdmin;
+import co.cask.cdap.data2.transaction.queue.QueueConstants;
+import co.cask.cdap.data2.transaction.queue.QueueEntryRow;
+import co.cask.cdap.data2.transaction.queue.QueueScanner;
+import co.cask.cdap.data2.util.TableId;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtilFactory;
+import co.cask.tephra.Transaction;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import co.cask.tephra.runtime.TransactionClientModule;
+import co.cask.tephra.runtime.TransactionModules;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.PeekingIterator;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.AbstractModule;
+import com.google.inject.Guice;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Singleton;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.twill.zookeeper.ZKClientService;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Debugging tool for queues in hbase.
+ */
+public class HBaseQueueDebugger extends AbstractIdleService {
+
+  private final HBaseQueueAdmin queueAdmin;
+  private final ZKClientService zkClientService;
+  private final HBaseQueueClientFactory queueClientFactory;
+  private final TransactionExecutorFactory txExecutorFactory;
+
+  @Inject
+  public HBaseQueueDebugger(HBaseQueueAdmin queueAdmin,
+                            HBaseQueueClientFactory queueClientFactory,
+                            ZKClientService zkClientService,
+                            TransactionExecutorFactory txExecutorFactory) {
+    this.queueAdmin = queueAdmin;
+    this.queueClientFactory = queueClientFactory;
+    this.zkClientService = zkClientService;
+    this.txExecutorFactory = txExecutorFactory;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    zkClientService.startAndWait();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    zkClientService.stopAndWait();
+  }
+
+  /**
+   * Only works for {@link co.cask.cdap.data2.transaction.queue.hbase.ShardedHBaseQueueStrategy}.
+   */
+  private void scanQueue(final QueueName queueName) throws Exception {
+    HBaseConsumerStateStore stateStore = queueAdmin.getConsumerStateStore(queueName);
+
+    TransactionExecutor txExecutor = Transactions.createTransactionExecutor(txExecutorFactory, stateStore);
+    txExecutor.execute(new TransactionExecutor.Procedure<HBaseConsumerStateStore>() {
```

Comment:
This will actually run with "short" transaction, which scanning queue can get a timeout. I notice that currently there is no way to run long running transaction with TransactionExecutor or transaction context. Can you file a JIRA for Tephra to add that support?